### PR TITLE
Add timeline & time column to calendar

### PR DESCRIPTION
**Pull request by asimmetti**
<comment on what this PR does>

closes #<issue number>

---

### Implementation

- [x] Does this code change accomplish what it is supposed to do?
- [ ] Can this solution be simplified without unreasonable effort
- [ ] Is the code modular enough?
- [ ] Does similar functionality already exist in the codebase? If yes, why isn’t it reused?
- [ ] Single Responsibility Principle: Does any class serve only one purpose? If not, it should probably split up in multiple classes.
- [ ] Interface Segregation: If interface implemented, is it fully implemented?
- [ ] Dependency Injection: Are central object instantiation processes managed by a single context?
- [ ] If any functionality is missing, is it marked with a comment that starts with TODO:?
- [ ] Do you see any potential to improve the performance of the code significantly? 

### Logic Errors and Bugs

- [ ] Can you think of any use case in which the code does not behave as intended?
- [ ] Can you think of any inputs or external events that could break the code?
- [ ] Are edge cases covered by tests?

### Error Handling and Logging

- [ ] Is error handling done the correct way?
- [ ] Should any logging or debugging information be added or removed?
- [ ] Are error messages user-friendly?

### Dependencies (internal and external)

- [ ] Were updates to documentation, configuration, or readme files made as required by this change?
- [ ] Is the API.md up-to-date?
- [ ] Are the enumerations between frontend and backend synchronized?

### Security and Data Privacy

- [ ] Does the code introduce any security vulnerabilities?
- [ ] Is (user) input validated, sanitized, and escaped to prevent security attacks such as cross-site scripting or SQL injection?
- [ ] Is sensitive data like user data handled and stored securely?
- [ ] Is the right encryption used?
- [ ] Is data retrieved from external APIs or libraries checked for security issues?
- [ ] Were any secrets committed?

### Testing and Testability (later)

- [ ] Have automated tests been added, or have related ones been updated to cover the change?
- [ ] Do the existing tests reasonably cover the code change?
- [ ] Are there some test cases, input or edge cases that should be tested in addition?

### Usability and Accessibility

- [ ] Is the proposed solution (UI) accessible (e.g. colorblindness, ...)?
- [x] Is the API/UI intuitive to use?
- [x] Was the app tested on another device?

### Readability

- [ ] Is the code easy to understand?
- [ ] Which parts were confusing to you and why?
- [ ] Can the readability of the code be improved by smaller methods/different names?
- [ ] Is the code located in the right file/folder?
- [ ] Do you think certain methods should be restructured to have a more intuitive control flow?
- [ ] Are there redundant or outdated comments?
- [ ] Could some comments convey the message better?
- [ ] Could some comments be removed by making the code itself more readable?

### CODE REVIEW

**Comment by JohaJung** [comment]
> [ ] Was the app tested on another device?

Ja, wurde auf Handy getestet, und dort ist das styling wirklich kaputt... es werden komischerweise zwei Tage untereinander angezeigt:

image



Auf dem Desktop sieht das schon ganz gut aus! Nur 

* warum nutzen wir nicht die volle Spaltenbreite?
* wenn zwei Veranstaltungen parallel liegen, sieht das ganz komisch aus

image

File: VLA-Frontend/src/components/calendar/DayColumn.tsx [hunk 1]
```diff
@@ -11,53 +9,63 @@ type DayColumnProps = {
...
-/**
- * DayColumn renders all events for a single day.
- * It is used inside the WeekGrid to build the calendar layout.
- */
-export default function DayColumn({ day, events, onEventClick, getEventColor }: DayColumnProps) {
-  return (
-  //TODO: Events rendern
 
-  // data-date helps with debugging (DevTools) and later for tests
+export default function DayColumn({
```
Comment (JohaJung) [inline VLA-Frontend/src/components/calendar/DayColumn.tsx]: This misses a docstring telling what this function does.

File: VLA-Frontend/src/components/calendar/Timeline.tsx [hunk 2]
```diff
@@ -0,0 +1,172 @@
...
+  if (!time) return null;
+  const parts = time.split(":");
+  if (parts.length < 2) return null;
+  const h = parseInt(parts[0], 10);
+  const m = parseInt(parts[1], 10);
+  if (Number.isNaN(h) || Number.isNaN(m)) return null;
+  return h * 60 + m;
+}
+
+function makeUnionFind(n: number) {
```
Comment (JohaJung) [inline VLA-Frontend/src/components/calendar/Timeline.tsx]: What does this function do? Please add a docstring.

File: VLA-Frontend/src/components/calendar/DayColumn.tsx [hunk 3]
```diff
@@ -11,53 +9,63 @@ type DayColumnProps = {
...
 
-        const eventProps = {
-          key: event.id,
-          className: `cv-event cv-event-${event.kind} ${getStatusClass(event.status)}`,
-          style: customColor ? { backgroundColor: customColor, borderColor: customColor } : undefined,
-          title: event.shortTitle || event.title,
-          ...(onEventClick && { onClick: () => onEventClick(event) }),
-        };
+        <div className="cv-dayMain">
+          {untimed.map((event) => {
```
Comment (JohaJung) [inline VLA-Frontend/src/components/calendar/DayColumn.tsx]: Ah, now I see why the right half of the column was empty: There, the events without time should be displayed!

But I propose to display them in one line directly below the Weekday name, and let the timeColumn start below that all-day-event-line

**Review by JohaJung — CHANGES_REQUESTED** [review]
UI-Wise on desktop, this looks already promising.

Sadly, I found the code hard to understand / read.

One fundamental question for my understanding: What are the three Layout components introduced here used for? I see

* TimeColumn
* Timeline
* TimelineColumn

And lost the overview which of those are used for what. Docstrings for those functions could probably beter my understanding :)

File: VLA-Frontend/src/styles/CalendarView.css [hunk 4]
```diff
@@ -1,15 +1,32 @@
...
+  --cv-day-count: 5;
+  /* Will be set inline from WeekGrid.tsx */
+  --cv-hour-count: 15;
+
+  --timelineRowBorderBottom: 1px dashed rgba(0, 0, 0, 0.4);
 }
-.cv-root {   
-  width: 100%;         
+
+/* ✅ IMPORTANT: avoid box model mismatches (line alignment bugs) */
```
Comment (JohaJung) [inline VLA-Frontend/src/styles/CalendarView.css]: Warum ein Emoji im Code? 👀

File: VLA-Frontend/src/components/calendar/CalendarView.tsx [hunk 5]
```diff
@@ -81,16 +78,15 @@ export default function CalendarView() {
...
+      <div className="cv-frame" style={frameStyleVars}>
         <WeekHeader days={days} />
         <WeekGrid
           days={days}
           eventsByDate={eventsByDate}
           onEventClick={handleEventClick}
         />
       </div>
-      {/* Modal overlay: create new event */}
+
```
Comment (JohaJung) [inline VLA-Frontend/src/components/calendar/CalendarView.tsx]: Why did you remove the Comments? Please re-add them if possible.

**Comment by MarcoLeipnitz** [comment]
This closes #91

File: VLA-Frontend/src/components/calendar/CalendarView.tsx [hunk 6]
```diff
@@ -56,41 +64,38 @@ export default function CalendarView() {
           type="button"
         />
 
-        <div className="cv-range" aria-label="Datumsbereich">
-          {rangeText}
-        </div>
+        <div className="cv-range">{rangeText}</div>
```
Comment (MarcoLeipnitz) [inline VLA-Frontend/src/components/calendar/CalendarView.tsx]: Why get rid of the aria-label?

File: VLA-Frontend/src/components/calendar/dateUtils.ts [hunk 7]
```diff
@@ -121,16 +120,61 @@ export function formatISODateDE(iso: string): string {
...
+/**
+ * Splits a datetime-local string (yyyy-mm-ddTHH:mm) into date and time parts.
+ * @param dateTimeString - Format: "2024-01-15T09:00"
+ * @returns Object with date (yyyy-mm-dd) and time (HH:mm)
+ */
+export function splitDateTime(dateTimeString: string): {
+  date: string;
+  time: string;
+} {
+  const [date, time] = dateTimeString.split("T");
```
[the remark this answers is not in the document]
Comment (MarcoLeipnitz) [reply VLA-Frontend/src/components/calendar/dateUtils.ts]: Look into formatting dates with an according formatter (Intl.DateTimeFormat)

File: VLA-Frontend/src/components/calendar/dateUtils.ts [hunk 8]
```diff
@@ -121,16 +120,61 @@ export function formatISODateDE(iso: string): string {
...
+  const date = new Date(dateTimeString);
+  date.setMinutes(date.getMinutes() + minutes);
+
+  const year = date.getFullYear();
+  const month = String(date.getMonth() + 1).padStart(2, "0");
+  const day = String(date.getDate()).padStart(2, "0");
+  const hours = String(date.getHours()).padStart(2, "0");
+  const mins = String(date.getMinutes()).padStart(2, "0");
+
+  return `${year}-${month}-${day}T${hours}:${mins}`;
```
Comment (MarcoLeipnitz) [inline VLA-Frontend/src/components/calendar/dateUtils.ts]: Add an @ return to the docstring, I would expect a date here

File: VLA-Frontend/src/components/calendar/DayColumn.tsx [hunk 9]
```diff
@@ -1,49 +1,150 @@
...
+   */
+  const timed: Appointment[] = [];
+  const untimed: Appointment[] = [];
+
+  /**
+   * 
+   * Classify events depending on whether they intersect
+   * the visible timeline window.
+   */
+  for (const e of events) {
```
Comment (MarcoLeipnitz) [inline VLA-Frontend/src/components/calendar/DayColumn.tsx]: Better name for e.
The for of loop should be avoided, it is hard to read in my opinion

File: VLA-Frontend/src/components/calendar/Timeline.tsx [hunk 10]
```diff
@@ -0,0 +1,204 @@
...
+           * Apply "short event" classes based on duration after clamping in CalenandarView.css:
+           *   cv-event-short     (< 90 min)
+           *   cv-event-veryShort (< 60 min)
+           */
+          const durationMin = Math.max(0, end - start);
+          const gutterPx = 6;
+          const widthPct = 100 / colCount;
+          const leftPct = col * widthPct;
+
+          const color = getEventColor?.(event) ?? event.series?.lecture?.color;
```
Comment (MarcoLeipnitz) [inline VLA-Frontend/src/components/calendar/Timeline.tsx]: Nice! I didn't know of the ?? operator

**Comment by MarcoLeipnitz** [comment]
Hello, I am a fellow branch intruder.
I have added a centralized file for managing the z-indexes as you implementation of the appointment also uses absolute positioned objects.
This fixes the first visual overlap bug @JohaJung mentioned

File: VLA-Frontend/src/components/calendar/Timeline.tsx [hunk 11]
```diff
@@ -0,0 +1,205 @@
...
+  getEventColor,
+}: Props) {
+  const minutesVisible = (endHour - startHour) * 60;
+
+  /**
+   * Convert events into numeric start/end minutes and clamp them to the visible window.
+   * This prevents events starting before 07:00 or ending after 22:00 from rendering outside.
+   */
+  const timed: TimedItem[] = events
+    .filter((e) => e.startTime && e.endTime)
```
Comment (MarcoLeipnitz) [inline VLA-Frontend/src/components/calendar/Timeline.tsx]: Again, this is always true

File: VLA-Frontend/src/components/calendar/Timeline.tsx [hunk 12]
```diff
@@ -0,0 +1,205 @@
...
+  colCount: number;
+};
+
+/**
+ * Layout events into columns for each overlap cluster.
+ * - Builds overlap clusters by time scanning
+ * - Assigns each event the first free column (greedy)
+ * - Uses the max simultaneous columns in that cluster as width divisor
+ */
+function layoutOverlaps(items: TimedItem[]): PositionedItem[] {
```
[the remark this answers is not in the document]
Comment (JohaJung) [reply VLA-Frontend/src/components/calendar/Timeline.tsx]: What do you mean?

**Comment by JohaJung** [comment]
> You cannot change whole day events without an internal server error

Can not reproduce. Can you give detailed steps to reproduce?

I would be surprised if this is an issue introduced by this branch. This branch only changes frontend layout stuff, but Internal server error would be triggered by backend.

File: VLA-Frontend/src/styles/zIndexes.css [hunk 13]
```diff
@@ -0,0 +1,18 @@
...
+    --cv-z-tooltip: 1200;
+    --cv-z-dayColumn-not-last-child-after: 20;
+    --cv-z-person-details-overlay: 1200 !important;
+
+    --z-dragcontainer: 999;
+    --z-dragcomponent: 998;
+
+    --z-global-note-container: 21;
+
+    --z-logger-toggle: 9999;
```
[the remark this answers is not in the document]
Comment (MarcoLeipnitz) [reply VLA-Frontend/src/styles/zIndexes.css]: This was an accident

File: VLA-Frontend/src/components/calendar/DayColumn.tsx [hunk 14]
```diff
@@ -1,47 +1,143 @@
...
+   
+    const startMin = e.startTime.getHours() * 60 + e.startTime.getMinutes();
```
[the remark this answers is not in the document]
Comment (JohaJung) [reply VLA-Frontend/src/components/calendar/DayColumn.tsx]: Fixed.
